File: IOS/Arrage Picture/Arrage Picture/Configuration.m
//
//  Configuration.m
//  100Numbers
//
//  Created by Nguyễn Thế Bình on 7/18/15.
//  Copyright (c) 2015 LapTrinhAlgo.Com. All rights reserved.
//

#import "Configuration.h"
#import "Define.h"
#import <GameKit/GameKit.h>

@implementation Configuration

-(void)ReportScore
{
    if(m_LeaderboardIdentifier != nil)
    {
        GKScore *score = [[GKScore alloc] initWithLeaderboardIdentifier:m_LeaderboardIdentifier];
        score.value = m_Score;
        
        [GKScore reportScores:@[score] withCompletionHandler:^(NSError *error)
         {
             if (error != nil)
             {
                 NSLog(@"%@", [error localizedDescription]);
             }
             else
             {
                 NSLog(@"Write game center succesful");
             }
         }];
    }
    else
    {
        NSLog(@"LeaderBoard failed");
    }
}

- (void) SetLeaderboardIdentifier : (NSString*) p_leaderboard
{
    m_LeaderboardIdentifier = p_leaderboard;
}

+(instancetype) GetSingleton
{
    static Configuration *ShareSingleton = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        ShareSingleton = [[Configuration alloc] init];
    });
    
    return ShareSingleton;
};

- (id)init
{
    self = [super init];
    if(self)
    {
        [self LoadConfig];
        m_LeaderboardIdentifier = nil;
    }
    
    return self;
}


- (void) LoadConfig
{
    NSString *pathData = [self GetPathData];
    NSLog(@"data.plist: %@", pathData);
    NSDictionary *dicData = [NSDictionary dictionaryWithContentsOfFile:pathData];
    if (dicData != nil)
    {
       
        m_IsMute = [dicData[@"IsMute"] boolValue];
        m_Level = [dicData[@"Level"] integerValue];
        m_TimePlay = [dicData[@"TimePlay"] integerValue];
        m_Score = [dicData[@"Score"] integerValue];
        if (m_Score == 0) {
            m_Score = MAXFLOAT;
        }

        NSLog(@"Level: %li and TimePlay: %li and Score: %li", m_Level, m_TimePlay, m_Score);
    }
    else
    {
        NSLog(@"Load data.plist fail !!");
        m_IsMute = FALSE;
        m_Level = 0;
        m_TimePlay = 0;
        m_Score = MAXFLOAT;
    }
}


- (BOOL) GetIsMute
{
    return m_IsMute;
}


- (NSInteger) GetLevel
{
    return m_Level;
}

- (void) WriteLevel:(NSInteger)p_level
{
    NSString *pathData = [self GetPathData];
    NSMutableDictionary *dicData = [NSMutableDictionary dictionaryWithContentsOfFile:pathData];
    if (dicData != nil)
    {
        
        [dicData setObject:[NSNumber numberWithInteger:p_level] forKey:@"Level"];
        [dicData writeToFile:pathData atomically:YES];
        m_Level = p_level;
        
        //NSLog(@"Score new: %li", (long)m_Score);
        
    }
    else
    {
        NSLog(@"Load data plist info fail !!");
    }
}


- (NSString*) GetPathData
{
    NSString *destinationPath = [[NSSearchPathForDirectoriesInDomains(
                                                                      NSDocumentDirectory,
                                                                      NSUserDomainMask, YES
                                                                      ) objectAtIndex:0] stringByAppendingString:FILECONFIG];
    
    // NSLog(@"File path: %@", destinationPath);
    
    if (![[NSFileManager defaultManager] fileExistsAtPath:destinationPath])
    {
        // The database file does not exist in the documents directory, so copy it from the main bundle now.
        NSString *sourcePath = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:FILECONFIG];
        NSError *error = nil;
        [[NSFileManager defaultManager] copyItemAtPath:sourcePath toPath:destinationPath error:&error];
        
        // Check if any error occurred during copying and display it.
        if (error != nil)
        {
            NSLog(@"%@", [error localizedDescription]);
        }
        
        NSLog(@"Copied data plist");
        NSLog(@"Source: %@", sourcePath);
        NSLog(@"Destination: %@", destinationPath);
    }
    else
    {
        NSLog(@"File exist");
    }
    
    return destinationPath;
    
}

- (void) WriteMute : (BOOL) p_ismute
{
    //NSLog(@"Writemute");
    
    NSString *pathData = [self GetPathData];
    NSMutableDictionary *dicData = [NSMutableDictionary dictionaryWithContentsOfFile:pathData];
    if (dicData != nil)
    {
        
        [dicData setObject:[NSNumber numberWithBool:p_ismute] forKey:@"IsMute"];
        [dicData writeToFile:pathData atomically:YES];
        
        m_IsMute = p_ismute;
        
    }
    else
    {
        NSLog(@"Load data plist info fail !!");
    }
    
};



- (UIImage*) TakeScreenshot
{
    CGSize imageSize = [[UIScreen mainScreen] bounds].size;
    if (NULL != UIGraphicsBeginImageContextWithOptions)
        UIGraphicsBeginImageContextWithOptions(imageSize, NO, 0);
    else
        UIGraphicsBeginImageContext(imageSize);
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // Iterate over every window from back to front
    for (UIWindow *window in [[UIApplication sharedApplication] windows])
    {
        if (![window respondsToSelector:@selector(screen)] || [window screen] == [UIScreen mainScreen])
        {
            // -renderInContext: renders in the coordinate space of the layer,
            // so we must first apply the layer's geometry to the graphics context
            CGContextSaveGState(context);
            // Center the context around the window's anchor point
            CGContextTranslateCTM(context, [window center].x, [window center].y);
            // Apply the window's transform about the anchor point
            CGContextConcatCTM(context, [window transform]);
            // Offset by the portion of the bounds left of and above the anchor point
            CGContextTranslateCTM(context,
                                  -[window bounds].size.width * [[window layer] anchorPoint].x,
                                  -[window bounds].size.height * [[window layer] anchorPoint].y);
            
            // Render the layer hierarchy to the current context
            [[window layer] renderInContext:context];
            
            // Restore the context
            CGContextRestoreGState(context);
        }
    }
    
    // Retrieve the screenshot image
    UIImage *screenImage = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    //NSData *imageDataForEmail = UIImageJPEGRepresentation(imageForEmail, 1.0);
    
    return screenImage;
    
}


- (NSInteger) GetTimePlay
{
    return m_TimePlay;
}


- (void) WriteNextTimePlay
{
    m_TimePlay ++;
    m_TimePlay = m_TimePlay >= 10000 ? 0 : m_TimePlay;
    NSString *pathData = [self GetPathData];
    NSMutableDictionary *dicData = [NSMutableDictionary dictionaryWithContentsOfFile:pathData];
    if (dicData != nil)
    {
        NSLog(@"Write timeplay: %li", m_TimePlay);
        [dicData setObject:[NSNumber numberWithInteger:m_TimePlay] forKey:@"TimePlay"];
        [dicData writeToFile:pathData atomically:YES];
    }
    else
    {
        NSLog(@"Load data plist info fail !!");
    }
}


- (NSInteger) GetScore
{
    return m_Score;
}

- (void) WriteScore : (NSInteger) p_score
{
    m_Score = m_Score > p_score ? p_score : m_Score;
    NSString *pathData = [self GetPathData];
    NSMutableDictionary *dicData = [NSMutableDictionary dictionaryWithContentsOfFile:pathData];
    if (dicData != nil)
    {
        NSLog(@"Write score: %li", m_Score);
        [dicData setObject:[NSNumber numberWithInteger:m_Score] forKey:@"Score"];
        [dicData writeToFile:pathData atomically:YES];
        
        [self ReportScore];
    }
    else
    {
        NSLog(@"Load data plist info fail !!");
    }
}

@end

File: IOS/PicturePuzzle/PicturePuzzle/Configuration.swift
//
//  Configuration.swift
//  CheckMoney
//
//  Created by Nguyễn Thế Bình on 12/28/15.
//  Copyright © 2015 Nguyễn Thế Bình. All rights reserved.
//
import UIKit
import GameKit

class Configuration
{
    static let ShareInstance = Configuration()
  
    
    private init()
    {
        LoadConfig()
        LoadListPhoto()
    }
  
    func CallSelf() -> Void
    {
        //print("Function do nothing")
    }
    
    func LoadConfig()
    {
        let pathData: String = GetDocPathFile(FILE_CONFIG)
        let dicData: NSDictionary? = NSDictionary(contentsOfFile: pathData)!
        if dicData != nil
        {
          if let l_ismuted = dicData?.objectForKey("ismuted") as? Bool
          {
            SoundController.ShareInstance.m_ismuted = l_ismuted
          }
          else
          {
            SoundController.ShareInstance.m_ismuted = false
          }
          
          if let l_level = dicData?.objectForKey("level") as? Int
          {
            PPCore.ShareInstance.m_level = l_level
          }
          else
          {
            PPCore.ShareInstance.m_level = 1
          }
          
          if let l_coin = dicData?.objectForKey("coin") as? Int
          {
            PPCore.ShareInstance.m_coin = l_coin
          }
          else
          {
            PPCore.ShareInstance.m_coin = 100
          }
          
        }
        else
        {
            NSLog("Load data.plist fail !!")
            SoundController.ShareInstance.m_ismuted = false
            PPCore.ShareInstance.m_level = 1
            PPCore.ShareInstance.m_coin = 100
        }
        
    }

    
    func WriteCoin(p_coin: Int)
    {
        let pathData: String = GetDocPathFile(FILE_CONFIG)
        let dicData: NSMutableDictionary? = NSMutableDictionary(contentsOfFile: pathData)!
        if let l_dicData = dicData
        {
            l_dicData.setObject(NSNumber(integer: p_coin), forKey:"coin")
            l_dicData.writeToFile(pathData, atomically:true)
        }
        else
        {
            NSLog("Load data plist info fail !!")
            
        }
    }
    
    func WriteMute(p_mute: Bool)
    {
        let pathData: String = GetDocPathFile(FILE_CONFIG)
        let dicData: NSMutableDictionary? = NSMutableDictionary(contentsOfFile: pathData)!
        if let l_dicData = dicData
        {
            l_dicData.setObject(p_mute, forKey:"ismuted")
            l_dicData.writeToFile(pathData, atomically:true)
        }
        else
        {
            NSLog("Load data plist info fail !!")
            
        }
    }
    
    //load list photo
    func LoadListPhoto() -> Void
    {
        let pathData: String = GetDocPathFile(FILE_DATA)
        let dicData: NSDictionary? = NSDictionary(contentsOfFile: pathData)!
        
        if dicData != nil
        {
            //print("Dic: \(dicData)")
            for (p_id, p_completed) in dicData!
            {
                //print("ID: \(p_id)")
                let l_photo = Photo(p_id: p_id as! String, p_completed: PHOTO_STATUS(rawValue: p_completed as! Int)!)
                PPCore.ShareInstance.m_ArrayPhoto.append(l_photo)
            }
            
            PPCore.ShareInstance.m_ArrayPhoto.sortInPlace{$0.m_id < $1.m_id}
        }
        else
        {
            NSLog("Load data.plist fail !!")
            
        }
    }
    
    func WriteComplete(p_level: Int, p_completed: Int)
    {
        let pathData: String = GetDocPathFile(FILE_DATA)
        let dicData: NSMutableDictionary? = NSMutableDictionary(contentsOfFile: pathData)!
        if let l_dicData = dicData
        {
            l_dicData.setObject(NSNumber(integer: p_completed), forKey: String(p_level))
            l_dicData.writeToFile(pathData, atomically:true)
        }
        else
        {
            NSLog("Load data plist info fail !!")
            
        }
    }

    
}

